### ISSUE TITLE
Remove all existing reliance on VSCodeVim remaps?

### ISSUE THREAD

**Issue by macintacos** [VSpaceCode/VSpaceCode#126]
## Problem

I've recently been interested in giving the [vscode-neovim](https://github.com/asvetliakov/vscode-neovim) extension a try, however my hesitancy stems from the fact that there are a few instances where VSpaceCode relies on VSCodeVim in order to perform certain actions. Only speaking to what's mapped in `package.json`, I can see the following instances:

- https://github.com/VSpaceCode/VSpaceCode/blob/ba040ddd1f35aa0dc5ed0e73d2eff0dc4b6b7fc9/package.json#L117-L128
- https://github.com/VSpaceCode/VSpaceCode/blob/ba040ddd1f35aa0dc5ed0e73d2eff0dc4b6b7fc9/package.json#L1084-L1096
- https://github.com/VSpaceCode/VSpaceCode/blob/ba040ddd1f35aa0dc5ed0e73d2eff0dc4b6b7fc9/package.json#L1097-L1113
- https://github.com/VSpaceCode/VSpaceCode/blob/ba040ddd1f35aa0dc5ed0e73d2eff0dc4b6b7fc9/package.json#L1120-L1135
- https://github.com/VSpaceCode/VSpaceCode/blob/ba040ddd1f35aa0dc5ed0e73d2eff0dc4b6b7fc9/package.json#L1417-L1429

I don't know if there are more instances in the codebase that would require tweaking based on this request.

## Solution

I have no idea if this is possible or not, but can we remove the reliance on VSCodeVim for these remappings, if only for the flexibility of using a different Vim extension by users? I know we can remap them, but I think at the end of the day this extension will be more robust if it has its own implementation (or calls existing VSCode commands directly) for the mappings it provides. Also, the barrier to entry for folks who enjoy the vscode-neovim extension will be lowered overall if this were done.

## Alternatives

I've considered remapping these commands myself (which I'll do if this issue doesn't make sense) however I wanted to see if this is a feasible request for the extension itself before I decided on a way forward

**Comment by stevenguh** [comment]
Related: https://github.com/VSpaceCode/VSpaceCode/issues/34#issuecomment-673235610

I am open to it but there is a few approaches I am thinking. In short-term, we should document, how users can use this extension with vscode-neovim or other vim provider. For example, document that trigger mechanism in the comment I linked, and overrides that change the bindings relied on VSCodeVim.

The longer term is to build a structure that can facilitate multiple vim provider.
1. Have a command proxy so commands like those that involve the same command on different vim provider. (btw, I can't find doc in vscode-neovim on how to involve vim command directly with vscode command)
2. Figure out how to onboard people if they are not using VSCodeVim. Since this is an extension pack that bundles VSCodeVim directly, it wouldn't be that great of an experience to have to uninstall VSCodeVim and install vscode-neovim and configure keybinding. In addition, our config commands only works for VSCodeVim at the moment. An idea is to have another extension that tailor specific for neovim and bundles the right extension while maintaining most of the current bindings. The structure might looks something like as follows, so the top level extension is only an extension pack, and all of functionality is offered by the core extension.
VSpaceCode --depends--> vspacecode-core --depends--> which-key
VSpaceCode-NeoVim --depends--> vspacecode-core --depends--> which-key

**Comment by macintacos** [comment]
I believe much of the configuration of "vim commands that call VSCode commands directly" happens in remappings that the user defines in their `init.vim` side of things in the case of the vscode-neovim extension, which goes something like this:

https://github.com/asvetliakov/vscode-neovim/blob/e61832119988bb1e73b81df72956878819426ce2/vim/vscode-code-actions.vim#L70

They _should_ probably document this better, which is probably something I'm going to contribute to that project pretty soon 🙂 

As far as onboarding... I'm not sure what the best approach is for that. If there was a way to decouple VSCodeVim from being _needed_ for all functions in the extension to work, but still provided  it as a default extension for an out-of-the-box "better" experience, that sounds like the best in my head.

Now that I think about it, perhaps this would be the difference between "evil" and "holy" mode when you're first setting up Spacemacs? The "Evil" mode will provide VSCodeVim as the default extension, and "Holy" will make it such that you don't have to have any Vim extension installed at all (but can install it or whatever other vim provider you want later if you wish)?

**Comment by stevenguh** [comment]
> I believe much of the configuration of "vim commands that call VSCode commands directly" happens in remappings that the user defines in their init.vim side of things in the case of the vscode-neovim extension, which goes something like this:
> 
> https://github.com/asvetliakov/vscode-neovim/blob/e61832119988bb1e73b81df72956878819426ce2/vim/vscode-code-actions.vim#L70
> 
> They should probably document this better, which is probably something I'm going to contribute to that project pretty soon 🙂

If I am not mistaken, this is the way neovim call VSCode command. I was lookin at a vscode command to call command in neovim (e.g. call `:noh` from the vscode side)

> As far as onboarding... I'm not sure what the best approach is for that. If there was a way to decouple VSCodeVim from being needed for all functions in the extension to work, but still provided it as a default extension for an out-of-the-box "better" experience, that sounds like the best in my head.

I am not sure either, that's why this is a bit longer term thing. The approach I described I above breaks the extension into two. However, some manual config is required to use vscode-neovim like setting the path and installation of neovim. Therefore, breaking into two extension *might* not make too much sense because we cannot automate all the installation upon installation (maybe that's fine?).

That leads to the other option similar to what you said evil/holy that everything will still be in the same extension. Upon installation of this extension, VSCode will be installed by default due it being in the extension pack config by VSCode. The user will need to manual uninstall Vim and set the vim provider in our config to vscode-neovim for example. Upon setting vim provider in the our config, we can prompt users to uninstall Vim and install vscode-neovim and checkout our doc for sure.

### Recap

Sort-term: Document the manual process for other vim users
Medium-term: Have a way to proxy vim command so they are not vendor specific
Long-term: Better on-board for non VSCodeVim user (may that be in the same extension or separate extension)

**Comment by macintacos** [comment]
Not sure if thisd helps, but I managed to run `:noh` via a command by setting up the following keymap for neovim:

```
  {
    "key": "ctrl+shift+tab",
    "command": "vscode-neovim.send",
    "when": "editorTextFocus && neovim.mode != insert",
    "args": ":noh<CR>"
  }
```

Hope that helps a little, since it definitely wasn't immediately obvious (sorry for misunderstanding before)

**Comment by stevenguh** [comment]
So an override like this should work?
```json
 { 
 	"keys": "s.c", 
 	"name": "Clear highlight", 
 	"type": "command", 
 	"command": "vscode-neovim.send", 
 	"args":  ":noh<CR>"
 }
```

If would be great if you can help us document the way to setup neovim in the short term :)

**Comment by macintacos** [comment]
@stevenguh just tested that, and at least for me it works exactly the way you'd expect!

I'm not entirely sure what else there is to document that isn't already explained in their wiki - you install the extension, ensure you point at the correct `nvim` binary, and it should "just work". As far as remapping things, calling `vscode-neovim.send` like we did above should have the same results as running a normal `:` command in neovim itself (since I think the extension just takes that command and forwards it to the neovim process in the background).

**Comment by stevenguh** [comment]
I don't think we will need to document much of the installation of the extension since we can just point to their wiki. I am thinking more on the latter, the customization needed on VSpaceCode config.

Rough outline/steps:
1. Uninstall VSCodeVim extension
2. Install vscode-neovim extension and following their installation guide
3. Remove vspacespace invocation in `vim.normalModeKeyBindingsNonRecursive` or `vim.visualModeKeyBindingsNonRecursive` of `settings.json`
4. Add vspacecode invocation to the vim config https://github.com/VSpaceCode/VSpaceCode/issues/34#issuecomment-673235610
5. Add overrides to convert the vscodevim specific call to neovim e.g. https://github.com/VSpaceCode/VSpaceCode/issues/126#issuecomment-711060596

**Comment by macintacos** [comment]
no. 3 in your list isn't strictly necessary - if you uninstall/disable VSCodeVim (step 1), the options for vim are just ignored. I'm switching between VSCodeVim/vscode-neovim daily just to get the hang of things and I don't have to edit any of my settings as long as both extensions aren't enabled at the same time.

**Comment by zetashift** [comment]
So is anybody already using neovim with VSpaceCode?

**Comment by macintacos** [comment]
I was for a time, but I've gone back to VSCodeVim (mostly due to multi-cursor stuff that is just handled better there at the moment).

**Comment by zetashift** [comment]
@macintacos how do I get started with vscode-neovim and VSpaceCode?

**Comment by stevenguh** [comment]
@zetashift The rough outline I described a while back should still work.

> Rough outline/steps:
> 1. Uninstall VSCodeVim extension
> 2. Install vscode-neovim extension and following their installation guide
> 3. Remove vspacespace invocation in `vim.normalModeKeyBindingsNonRecursive` or `vim.visualModeKeyBindingsNonRecursive` of `settings.json`
> 4. Add vspacecode invocation to the vim config https://github.com/VSpaceCode/VSpaceCode/issues/34#issuecomment-673235610
> 5. Add overrides to convert the vscodevim specific call to neovim e.g. https://github.com/VSpaceCode/VSpaceCode/issues/126#issuecomment-711060596

**Comment by Fraznist** [comment]
I am also interested in using vspacecode and vscode-neovim, and for the most part, it work once I add this to my `init.vim`
```
if exists('g:vscode')
        nnoremap <space> :call VSCodeNotify('whichkey.show')<CR>
endif
```
However, ~~disabling VsCodeVim~~ enabling vscode-neovim removes some of the top level menu items from VSpaceCode. So far I have found that the following menu items disappear:
- Compile/Comments (c)
- Major Mode (m), at least for python
- Jump/Join/Split

Adding the menu items manually to `settings.json`, within `vspacecode.bindingOverrides` does not add any menu items. In fact, `vspacecode.bindingOverrides` fails to add any bindings when vscode-neovim is enabled, neither entire menus, nor single command editions.

Does anybody else have the same issue of missing menu items? If menu items stayed, or could be overwritten, we could implement all VsCodeVim dependent features by mapping it through vscode-neovim via `vscode-neovim.send`

**Comment by stevenguh** [comment]
I am surprised to hear the menu items disappear. I expect that might not work but it should appear on the UI nonetheless. Without knowing your `settings.json` , I would assume there is some syntax error in your `vspacecode.bindingOverrides`.

You should able to override the menu item that uses VscodeVim command to `vscode-neovim.send` via `vspacecode.bindingOverrides`.

**Comment by Fraznist** [comment]
I am very surprised at the items missing as well. My `bindingOverrides` seems to be valid, it works when `vscode-neovim` is disabled. When `vscode-neovim` is enabled, the behavior is quite janky. For instance:
* My custom binding for keys `t.d` stops showing up.
* The same binding with all other fields same shows up if its keys are `s.c`
* The command and when clauses are vscode builtins.
Both top menu items, and individual commands disappear arbitrarily.
In case anyone wants to check my `settings.json`: [link](https://gist.github.com/Fraznist/e087d0ef3c6f926d0b11349200dd7988)
My current `bindingOverrides` works with `VsCodeVim`:
```json
  "vspacecode.bindingOverrides": [
    {
      "keys": "t.d",
      "name": "Toggle Peek Docstring",
      "type": "command",
      "command": "editor.action.showHover",
      "when": "editorTextFocus"
    }
  ]
```

Anyways, I don't want to hog this issue with bugs that only I experience, but I wanted to share my case since the topic came up. I will try to resolve my bug myself. I would appreciate hints though, such as where I could find any error logs vspacecode would log.

**Comment by stevenguh** [comment]
That's weird. I don't see anything weird except the `when` key value you used in the overrides has no meaning. Can you create another GitHub issue If you are still experiencing this issue? Let's look into this.

**Comment by mkcode** [comment]
Hi All, Quick message for anyone who comes across this - I successfully switched over to using VSCode-neovim. Much more performant! and surprising few rough edges.

I created a basic neovim config that works great with VSpaceCode here with instructions here: https://gist.github.com/mkcode/4d8077ddf473d0f4090607637a9b0493
